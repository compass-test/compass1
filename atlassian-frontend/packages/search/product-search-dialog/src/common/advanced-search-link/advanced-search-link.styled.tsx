import styled from 'styled-components';
import { gridSize } from '@atlaskit/theme/constants';
import { N900, N20, N0, B75, B100 } from '@atlaskit/theme/colors';
import { Link, LinkComponentProps } from '@atlassian/search-dialog';
import { themed } from '@atlaskit/theme/components';

const outlineColor = themed({ light: B100, dark: B75 });

export const CenteredAdvancedSearchGroup = styled.div`
  display: flex;
  align-items: center;
`;

const outlineWidth = 2;
const outlineRemovedFromPaddingWidth = outlineWidth / 2;

export const StyledAdvancedSearch = styled(Link)<
  LinkComponentProps & {
    onClick?: (e: React.MouseEvent | KeyboardEvent) => any;
  }
>`
  align-items: center;
  color: ${N900};
  cursor: pointer;
  display: flex;
  width: 100%;
  justify-content: space-between;
  padding: ${gridSize() - outlineRemovedFromPaddingWidth}px
    ${2 * gridSize() - outlineRemovedFromPaddingWidth}px;

  border: ${outlineWidth}px solid transparent;

  &:hover {
    color: ${N900};
    background-color: ${N20};
    text-decoration: none;
  }

  :visited {
    color: ${N900};
  }

  :focus {
    outline: none;
    border: ${outlineWidth}px solid ${outlineColor};
  }

  background-color: ${(props) => (props.isKeyboardHighlighted ? N20 : N0)};
`;

export const AdvancedSearchContent = styled.span`
  font-size: 14px;
  line-height: ${gridSize() * 2}px;
  padding-left: ${gridSize() * 2}px;
`;
