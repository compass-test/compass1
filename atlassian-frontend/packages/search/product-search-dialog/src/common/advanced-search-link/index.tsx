export { AdvancedSearchLink } from './advanced-search-link';
export {
  AdvancedSearchContent,
  CenteredAdvancedSearchGroup,
  StyledAdvancedSearch,
} from './advanced-search-link.styled';

export { default as MultiSiteAdvancedSearchLink } from './multi-site-advanced-search-link';
