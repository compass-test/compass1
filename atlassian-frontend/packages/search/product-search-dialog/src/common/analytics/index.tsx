import * as ContextImports from './context';
export type SupportedContext = ContextImports.SupportedContext;
export type AnalyticsContextAction = ContextImports.AnalyticsContextAction;
export { SearchDialogAnalyticsContext } from './search-dialog-analytics-context';
export { SearchResultsShownHandler } from './results-shown-handler';
export { withAnalytics } from './with-analytics';
export { useAnalytics } from './use-analytics';
export { DialogDismissedHandler } from './dialog-dismissed-handler';
export { TextEnteredHandler } from './text-entered-handler';
export { useAnalyticsContext } from './context';
export { sha1Hash, getTrigger, isNewTab, getWordCount } from './utils';
export {
  onFilterUnselect,
  onFilterSelect,
  onAdvancedSearchSelected,
  onMultiSiteAdvancedSearchSelected,
  onDismissed,
  onExperimentLoaded,
  onFiltersCleared,
  onMoreFiltersSelected,
  onPaneLoadComplete,
  onPreQueryEmpty,
  onPreQueryScreenViewed,
  onRequestMade,
  onSearchResultHighlighted,
  onSearchResultSelected,
  onSearchResultsShown,
  onShowAllClicked,
  onShowMoreClicked,
  onTabSelected,
  onTextEntered,
  AdvancedSearchLinkSubjectId,
  selectEvent,
} from './events';

export type {
  ScreenType,
  NonPrivacySafeContext,
  OnPreQueryScreenViewedProps,
  OnSearchResultHighlightedProps,
  OnSearchResultSelectedProps,
  OnShowAllClickedProps,
  OnShowMoreClickedProps,
  SearchResultContextPayload,
  SectionID,
  ResultAnalyticData,
  ResultSectionAnalyticData,
} from './events';
