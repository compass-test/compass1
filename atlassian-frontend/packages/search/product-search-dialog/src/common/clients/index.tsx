import * as CommonTypes from './common-types';

export { AggregatorClient } from './base-search-client';
export { CollaborationGraphClient } from './collaboration-graph-client';
export { responseErrorToError } from './error-handler';
export { extractSiteUrl } from './multi-site-utils';
export { timed } from './timing';

export type ModelParam = CommonTypes.ModelParam;
export type SearchContext = CommonTypes.SearchContext;
export type AggregatorConfig = import('./base-search-client').AggregatorConfig;
export type ProductsPermissionsResponse<
  T
> = import('./base-search-client').ProductsPermissionsResponse<T>;
export type CollaborationGraphConfig = import('./collaboration-graph-client').CollaborationGraphConfig;
export type ABTest = CommonTypes.ABTest;
export type ContainerType = CommonTypes.ContainerType;
export type ScopedAggregatorResponse<T> = CommonTypes.ScopedAggregatorResponse<
  T
>;
export type CollaborationGraphResponse<
  T
> = CommonTypes.CollaborationGraphResponse<T>;
export type CollaborationGraphContainer = CommonTypes.CollaborationGraphContainer;
export type SharedClient = CommonTypes.SharedClient;

// Multi site
export type Site = CommonTypes.Site;
