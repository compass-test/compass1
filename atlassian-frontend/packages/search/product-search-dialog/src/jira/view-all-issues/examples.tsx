import React from 'react';
import { IntlProvider } from 'react-intl';
import ViewAllIssuesLink from './view-all-issues';
import { action } from '@storybook/addon-actions';

export const basic = () => (
  <IntlProvider locale="en">
    <ViewAllIssuesLink query="test" onClick={action('onAdvancedSearchClick')} />
  </IntlProvider>
);

export default {
  title: 'Jira Search Dialog/Search Results/Advanced issue search link',
};
