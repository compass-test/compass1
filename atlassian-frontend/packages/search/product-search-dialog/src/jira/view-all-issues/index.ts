export { default as ViewAllIssuesLink } from './view-all-issues';
