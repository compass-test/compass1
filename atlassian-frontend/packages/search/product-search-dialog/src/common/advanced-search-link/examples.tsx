import React from 'react';
import { SearchDialog, SearchFooter } from '@atlassian/search-dialog';
import { action } from '@storybook/addon-actions';
import { InjectedIntlProps, IntlProvider } from 'react-intl';
import { Products } from '../../common/product-context';
import { generateMockSitesWithAdvancedSearch } from '../../utils/site-util';
import {
  MultiSiteAdvancedSearchLink,
  Props,
} from './multi-site-advanced-search-link';

const commonProps: Props = {
  sites: [
    ...generateMockSitesWithAdvancedSearch(5, Products.confluence),
    ...generateMockSitesWithAdvancedSearch(5, Products.jira),
  ],
  isLoading: false,
  onClick: (href, e) => {
    e.preventDefault();
    return action(`onclick ${href}`)(e);
  },
};

export const MultiSite = (props: InjectedIntlProps) => (
  <MultiSiteAdvancedSearchLink {...props} {...commonProps} />
);

export const ManySites = (props: InjectedIntlProps) => (
  <MultiSiteAdvancedSearchLink
    {...props}
    {...commonProps}
    sites={generateMockSitesWithAdvancedSearch(100, Products.jira)}
  />
);

export const NoDropdown = (props: InjectedIntlProps) => (
  <MultiSiteAdvancedSearchLink
    {...props}
    {...commonProps}
    sites={generateMockSitesWithAdvancedSearch(4, Products.jira)}
  />
);

export default {
  title: 'common/Advanced Search Link',
  decorators: [
    (story: () => React.ElementType) => (
      <IntlProvider locale="en">
        <SearchDialog>
          <SearchFooter>{story()}</SearchFooter>
        </SearchDialog>
      </IntlProvider>
    ),
  ],
};
