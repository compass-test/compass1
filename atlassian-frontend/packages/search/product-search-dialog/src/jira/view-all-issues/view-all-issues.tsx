import React, { useCallback } from 'react';
import SearchIcon from '@atlaskit/icon/glyph/search';
import {
  ReturnIcon,
  useKeyboardNavigation,
  LinkComponent,
} from '@atlassian/search-dialog';
import {
  AdvancedSearchLink,
  AdvancedSearchContent,
  CenteredAdvancedSearchGroup,
} from '../../common/advanced-search-link';
import { messages } from '../../messages';
import { FormattedMessage, injectIntl, InjectedIntlProps } from 'react-intl';
import { getTrigger, isNewTab, withAnalytics } from '../../common/analytics';
import {
  AdvancedSearchLinkSubjectId,
  AdvancedSearchAnalyticsProps,
  Trigger,
  onAdvancedSearchSelected,
} from '../../common/analytics/events';
import { useFiltersAdvancedSearchUrlFactory } from '../jira-advanced-search-url-factory';
import { useJiraSearchClientContext } from '../clients';

interface Props {
  query: string;
  linkComponent?: LinkComponent;
  advancedSearchSelected: (props: AdvancedSearchAnalyticsProps) => any;
  onClick: (href: string, event: React.MouseEvent | KeyboardEvent) => void;
}

const onAdvancedSearchClicked = (
  e: React.MouseEvent<HTMLSpanElement, MouseEvent>,
  props: Props,
  href: string,
) => {
  const { advancedSearchSelected, onClick } = props;
  advancedSearchSelected({
    trigger: getTrigger(e),
    actionSubjectId: AdvancedSearchLinkSubjectId.JIRA_ISSUES,
    isLoading: false,
    newTab: isNewTab(e),
  });
  onClick(href, e);
};

const onKeyDown = (
  e: KeyboardEvent,
  href: string,
  advancedSearchSelected: (props: AdvancedSearchAnalyticsProps) => any,
  onClick: (href: string, event: React.MouseEvent | KeyboardEvent) => void,
) => {
  if (e.key === 'Enter' && !e.shiftKey) {
    e.preventDefault();
    advancedSearchSelected({
      trigger: Trigger.RETURN,
      actionSubjectId: AdvancedSearchLinkSubjectId.JIRA_ISSUES,
      isLoading: false,
      newTab: false,
    });
    onClick(href, e);
  }
};

function ViewAllIssuesLink(props: Props & InjectedIntlProps) {
  const { query, linkComponent, advancedSearchSelected, intl, onClick } = props;
  const { siteUrl } = useJiraSearchClientContext();
  const href = useFiltersAdvancedSearchUrlFactory()(query, siteUrl);

  const [isKeyboardHighlighted, ref] = useKeyboardNavigation<HTMLSpanElement>({
    onKeydownCallback: useCallback(
      (e) => onKeyDown(e, href, advancedSearchSelected, onClick),
      [advancedSearchSelected, href, onClick],
    ),
  });

  return (
    <span
      ref={ref}
      data-test-id={'jira-view-all-issues'}
      style={{ width: '100%', display: 'flex' }}
    >
      <AdvancedSearchLink
        onClick={(e: any) => onAdvancedSearchClicked(e, props, href)}
        href={href}
        isKeyboardHighlighted={isKeyboardHighlighted}
        linkComponent={linkComponent}
      >
        <CenteredAdvancedSearchGroup>
          <SearchIcon
            size="small"
            label={intl.formatMessage(messages.jira_view_all_issues)}
          />
          <AdvancedSearchContent>
            <FormattedMessage {...messages.jira_view_all_issues} />
          </AdvancedSearchContent>
        </CenteredAdvancedSearchGroup>
        <ReturnIcon inverted={isKeyboardHighlighted} />
      </AdvancedSearchLink>
    </span>
  );
}

export const ViewAllIssuesLinkWithIntl = injectIntl(ViewAllIssuesLink);

export default withAnalytics({
  advancedSearchSelected: onAdvancedSearchSelected,
})(injectIntl(ViewAllIssuesLink));
