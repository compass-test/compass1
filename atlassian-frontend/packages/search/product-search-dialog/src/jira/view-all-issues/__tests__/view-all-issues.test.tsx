import React from 'react';
import { mountWithIntl } from '../../../__tests__/__fixtures__/intl-test-helpers';
import { ViewAllIssuesLinkWithIntl as ViewAllIssuesLink } from '../view-all-issues';
import { AdvancedSearchLink } from '../../../common/advanced-search-link';
import { useKeyboardNavigation } from '@atlassian/search-dialog';

jest.mock('../../../common/advanced-search-link', () => ({
  AdvancedSearchContent: () => 'div',
  AdvancedSearchLink: () => 'div',
  CenteredAdvancedSearchGroup: () => 'div',
}));

let mockSiteUrl = {
  siteUrl: '',
};

jest.mock('../../clients', () => ({
  useJiraSearchClientContext: () => mockSiteUrl,
}));

jest.mock('@atlassian/search-dialog', () =>
  Object.assign({}, jest.requireActual('@atlassian/search-dialog'), {
    useKeyboardNavigation: jest.fn(),
    ReturnIcon: () => <div />,
    LinkComponent: () => <div />,
  }),
);

const advancedSearchSelected = jest.fn();
const onClick = jest.fn();
const searchSessionId = 'testSessionId';
const isLoading = false;

const commonProps = {
  searchSessionId,
  advancedSearchSelected,
  isLoading,
  onClick,
};

describe('<ViewAllIssuesLink />', () => {
  const locationAssignSpy = jest.fn().mockImplementation(() => null);
  const originalLocation = window.location;

  beforeAll(() => {
    Object.defineProperty(window, 'location', {
      value: {
        assign: locationAssignSpy,
      },
    });
  });

  afterAll(() => {
    window.location = originalLocation;
  });

  beforeEach(() => {
    jest.clearAllMocks();
    (useKeyboardNavigation as jest.Mock).mockReturnValue([false]);
    mockSiteUrl = {
      siteUrl: '',
    };
  });

  it('snapshot default', () => {
    const wrapper = mountWithIntl(
      <ViewAllIssuesLink {...commonProps} query={''} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('snapshot keyboard highlighted', () => {
    (useKeyboardNavigation as jest.Mock).mockReturnValue([true]);
    const wrapper = mountWithIntl(
      <ViewAllIssuesLink {...commonProps} query={''} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render AdvancedSearchLink with query string attached to href', () => {
    const wrapper = mountWithIntl(
      <ViewAllIssuesLink {...commonProps} query={'A test search'} />,
    );
    const advancedSearchLink = wrapper.find(AdvancedSearchLink).prop('href');
    expect(advancedSearchLink).toEqual(
      '/secure/QuickSearch.jspa?searchString=A%20test%20search',
    );
  });

  it('should render AdvancedSearchLink with site url and query string attached to href', () => {
    mockSiteUrl = {
      siteUrl: 'SOME_SITE',
    };

    const wrapper = mountWithIntl(
      <ViewAllIssuesLink {...commonProps} query={'A test search'} />,
    );

    const advancedSearchLink = wrapper.find(AdvancedSearchLink).prop('href');
    expect(advancedSearchLink).toEqual(
      'SOME_SITE/secure/QuickSearch.jspa?searchString=A%20test%20search',
    );
  });

  it('on enter should call onClick handler with url and event', () => {
    mountWithIntl(
      <ViewAllIssuesLink {...commonProps} query={'A test search'} />,
    );

    const event = new KeyboardEvent('keydown', { key: 'Enter' });
    (useKeyboardNavigation as jest.Mock).mock.calls[0][0]['onKeydownCallback'](
      event,
    );

    expect(onClick).toBeCalledWith(
      '/secure/QuickSearch.jspa?searchString=A%20test%20search',
      event,
    );
  });

  describe('analytics', () => {
    it('on click with no modifier should trigger analytics event with newTab false', () => {
      const wrapper = mountWithIntl(
        <ViewAllIssuesLink {...commonProps} query={'A test search'} />,
      );

      wrapper.find(AdvancedSearchLink).prop('onClick')({} as React.MouseEvent);

      expect(advancedSearchSelected).toBeCalledWith({
        trigger: 'click',
        actionSubjectId: 'jiraIssuesSearchLink',
        isLoading,
        newTab: false,
      });
    });

    ['ctrlKey', 'shiftKey', 'metaKey'].forEach((modifier) => {
      it(`on click with ${modifier} should trigger analytics event with newTab true`, () => {
        const wrapper = mountWithIntl(
          <ViewAllIssuesLink {...commonProps} query={'A test search'} />,
        );

        wrapper.find(AdvancedSearchLink).prop('onClick')({
          [modifier]: true,
        } as any);

        expect(advancedSearchSelected).toBeCalledWith({
          trigger: 'click',
          actionSubjectId: 'jiraIssuesSearchLink',
          isLoading,
          newTab: true,
        });
      });
    });

    it('on enter should trigger analytic events', () => {
      mountWithIntl(
        <ViewAllIssuesLink {...commonProps} query={'A test search'} />,
      );

      (useKeyboardNavigation as jest.Mock).mock.calls[0][0][
        'onKeydownCallback'
      ](new KeyboardEvent('keydown', { key: 'Enter' }));

      expect(advancedSearchSelected).toBeCalledWith({
        trigger: 'return',
        actionSubjectId: 'jiraIssuesSearchLink',
        isLoading,
        newTab: false,
      });
    });
  });
});
