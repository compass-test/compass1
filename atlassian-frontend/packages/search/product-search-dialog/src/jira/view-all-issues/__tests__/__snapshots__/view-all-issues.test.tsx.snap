// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewAllIssuesLink /> snapshot default 1`] = `
<ViewAllIssuesLink
  advancedSearchSelected={[MockFunction]}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  isLoading={false}
  onClick={[MockFunction]}
  query=""
  searchSessionId="testSessionId"
>
  <span
    data-test-id="jira-view-all-issues"
    style={
      Object {
        "display": "flex",
        "width": "100%",
      }
    }
  >
    <AdvancedSearchLink
      href="/secure/QuickSearch.jspa?searchString="
      isKeyboardHighlighted={false}
      onClick={[Function]}
    >
      div
    </AdvancedSearchLink>
  </span>
</ViewAllIssuesLink>
`;

exports[`<ViewAllIssuesLink /> snapshot keyboard highlighted 1`] = `
<ViewAllIssuesLink
  advancedSearchSelected={[MockFunction]}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  isLoading={false}
  onClick={[MockFunction]}
  query=""
  searchSessionId="testSessionId"
>
  <span
    data-test-id="jira-view-all-issues"
    style={
      Object {
        "display": "flex",
        "width": "100%",
      }
    }
  >
    <AdvancedSearchLink
      href="/secure/QuickSearch.jspa?searchString="
      isKeyboardHighlighted={true}
      onClick={[Function]}
    >
      div
    </AdvancedSearchLink>
  </span>
</ViewAllIssuesLink>
`;
